### PR TITLE
[Teaser] Add support for multiple call to action (CTA) (#155)

### PR DESCRIPTION
**Pull request by vladbailescu**
First stage of changes to support CTAs:

* Removed link text
* Added fallback values for title and description
* Updated tests

### CODE REVIEW

File: content/src/content/jcr_root/apps/core/wcm/sandbox/components/teaser/v1/teaser/teaser.html [hunk 1]
```diff
@@ -20,9 +20,11 @@
 
     <div class="cmp-teaser__image" data-sly-resource="${teaser.imageResource}"></div>
 
-    <h2 class="cmp-teaser__title" data-sly-test.title="${teaser.title}">${title}</h2>
-    <h3 class="cmp-teaser__description" data-sly-test.description="${teaser.description}">${description}</h3>
-
-    <a class="cmp-teaser__link" href="${teaser.linkURL}">${teaser.linkText}</a>
```
Comment (godanny86) [inline content/src/content/jcr_root/apps/core/wcm/sandbox/components/teaser/v1/teaser/teaser.html]: What is the reasoning behind removing the anchor 'cmp-teaser__link' element completely? Some customers would appreciate a separate anchor tag that can be configured to look like a Call To Action button i.e: https://getbootstrap.com/docs/4.0/components/card/ ?

File: content/src/content/jcr_root/apps/core/wcm/sandbox/components/teaser/v1/teaser/teaser.html [hunk 2]
```diff
@@ -20,9 +20,11 @@
...
+    <h2 class="cmp-teaser__title" data-sly-test.title="${teaser.title}">
+        <a class="cmp-teaser__link" href="${teaser.linkURL}" data-sly-unwrap="${teaser.hideTitleLink}">${title}</a>
```
Comment (davidjgonzalez) [inline content/src/content/jcr_root/apps/core/wcm/sandbox/components/teaser/v1/teaser/teaser.html]: The repeated use of `cmp-teaser__link` seems off in the context of BEM as it requires nesting for context (else it cannot be differentiated from the descriptions link). Would this be better named `cmp-teaster__title-link` ?

File: content/src/content/jcr_root/apps/core/wcm/sandbox/components/teaser/v1/teaser/teaser.html (hunk 1 above)
Comment (vladbailescu) [reply content/src/content/jcr_root/apps/core/wcm/sandbox/components/teaser/v1/teaser/teaser.html]: @godanny86 We plan to add the possibility of having CTAs as a second stage, in a few days.

File: content/src/content/jcr_root/apps/core/wcm/sandbox/components/teaser/v1/teaser/teaser.html (hunk 2 above)
Comment (vladbailescu) [reply content/src/content/jcr_root/apps/core/wcm/sandbox/components/teaser/v1/teaser/teaser.html]: Thanks for the suggestion, it makes a lot of sense

File: content/src/content/jcr_root/apps/core/wcm/sandbox/components/teaser/v1/teaser/teaser.html (hunk 1 above)
Comment (godanny86) [reply content/src/content/jcr_root/apps/core/wcm/sandbox/components/teaser/v1/teaser/teaser.html]: great!

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/sandbox/models/Teaser.java [hunk 3]
```diff
@@ -24,70 +24,130 @@
...
+    /**
+     * Checks if link should be hidden on the image.
      *
-     * @return the teaser's description or {@code null}
-     * @since com.adobe.cq.wcm.core.components.sandbox.models 2.3.0
+     * @return {@code true} if link should be hidden on the image, {@code false} otherwise
+     * @since com.adobe.cq.wcm.core.components.sandbox.models 1.0.0
      */
-    default String getDescription() {
+    default boolean isHideImageLink() {
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/sandbox/models/Teaser.java]: I'd call it `hideImageLink`

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/sandbox/models/Teaser.java [hunk 4]
```diff
@@ -24,70 +24,130 @@
...
+     * Checks if link should be hidden on the title.
      *
-     * @return the text of the URL to which teaser links or {@code null}
-     * @see #getLinkURL()
-     * @since com.adobe.cq.wcm.core.components.sandbox.models 2.3.0
+     * @return {@code true} if link should be hidden on the title, {@code false} otherwise
+     * @since com.adobe.cq.wcm.core.components.sandbox.models 1.0.0
      */
-    default String getLinkText() {
+    default boolean isHideTitleLink() {
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/sandbox/models/Teaser.java]: I'd call it `hideTitleLink`

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/sandbox/models/Teaser.java [hunk 5]
```diff
@@ -24,70 +24,130 @@
...
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Checks if link should be hidden on the description.
+     *
+     * @return {@code true} if link should be hidden on the description, {@code false} otherwise
+     * @since com.adobe.cq.wcm.core.components.sandbox.models 1.0.0
+     */
+    default boolean isHideDescriptionLink() {
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/sandbox/models/Teaser.java]: I'd call it `isHideDescriptionLink`.

File: content/src/content/jcr_root/apps/core/wcm/sandbox/components/teaser/v1/teaser/README.md [hunk 6]
```diff
@@ -44,8 +44,9 @@ The following properties are written to JCR for this Teaser component and are ex
 BLOCK cmp-teaser
     ELEMENT cmp-teaser__image
     ELEMENT cmp-teaser__title
+    ELEMENT cmp-teaser__title-link
```
Comment (jckautzmann) [inline content/src/content/jcr_root/apps/core/wcm/sandbox/components/teaser/v1/teaser/README.md]: I'd call it `cmp-teaser__link-title`

File: content/src/content/jcr_root/apps/core/wcm/sandbox/components/teaser/v1/teaser/README.md [hunk 7]
```diff
@@ -44,8 +44,9 @@ The following properties are written to JCR for this Teaser component and are ex
...
     ELEMENT cmp-teaser__description
-    ELEMENT cmp-teaser__link
+    ELEMENT cmp-teaser__description-link
```
Comment (jckautzmann) [inline content/src/content/jcr_root/apps/core/wcm/sandbox/components/teaser/v1/teaser/README.md]: I'd call it `cmp-teaser__link-description`.

File: content/src/content/jcr_root/apps/core/wcm/sandbox/components/teaser/v1/teaser/_cq_dialog/.content.xml [hunk 8]
```diff
@@ -1,66 +1,96 @@
...
+                            <file
+                                    jcr:primaryType="nt:unstructured"
+                                    sling:resourceType="cq/gui/components/authoring/dialog/fileupload"
+                                    allowUpload="{Boolean}false"
+                                    autoStart="{Boolean}false"
+                                    class="cq-droptarget"
+                                    fieldLabel="Image asset"
+                                    fileNameParameter="./fileName"
+                                    fileReferenceParameter="./fileReference"
+                                    mimeTypes="[image/gif,image/jpeg,image/png,image/webp,image/tiff]"
```
Comment (jckautzmann) [inline content/src/content/jcr_root/apps/core/wcm/sandbox/components/teaser/v1/teaser/_cq_dialog/.content.xml]: I'd add `image/svg+xml` as we will support SVG images.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/sandbox/models/Teaser.java (hunk 3 above)
Comment (gknobloch) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/sandbox/models/Teaser.java]: I prefer @vladbailescu 's approach, `is...` is a convention for `boolean`
`hideImageLink` sounds more like an action

Comment (jckautzmann) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/sandbox/models/Teaser.java]: it's just that isHide... does not sound well. But I guess that the convention `is<property-name>` is fine. We could also rename the property `imageLinkHidden`. The method name `isImageLinkHidden` would then sound better. I'm fine with both.

**Comment by vladbailescu** [comment]
@jckautzmann, thanks! I don't have strong feelings for the names, so I can change them if you feel it makes better sense to have `xLinkHidden` instead of `hideXLink`. For me it made better sense to have the common `hide` prefix as they would be grouped together nicely in CRX/DE ;)

**Pull request merged**